src/storage/store/engine.rs

```diff
@@ -813,6 +831,118 @@ impl ShardEngine {
         Ok((account_root, events, validation_errors))
     }
 
+    /// Checks if a removed verification was set as the user's primary address and revokes it if necessary.
+    ///
+    /// When a verification is removed from a user's account, this function ensures that if the
+    /// removed address was also set as their primary address (for either Ethereum or Solana),
+    /// the primary address setting is also revoked to maintain consistency.
+    ///
+    fn check_and_revoke_primary_address(
+        &mut self,
+        fid: u64,
+        deleted_verification: &proto::VerificationAddAddressBody,
+        txn_batch: &mut RocksDbTransactionBatch,
+    ) -> Result<(), MessageValidationError> {
+        // Determine protocol from removal message
+        let protocol = match deleted_verification.protocol {
+            x if x == proto::Protocol::Ethereum as i32 => proto::Protocol::Ethereum,
+            x if x == proto::Protocol::Solana as i32 => proto::Protocol::Solana,
+            _ => return Err(MessageValidationError::AddressNotPartOfVerification),
+        };
+
+        // Parse address based on protocol
+        let parsed_address = match protocol {
+            proto::Protocol::Ethereum => {
+                Address::from_slice(&deleted_verification.address).to_checksum(None)
+            }
+            proto::Protocol::Solana => bs58::encode(&deleted_verification.address).into_string(),
+        };
+
+        // Determine user data type based on protocol
+        let user_data_type = match protocol {
+            proto::Protocol::Ethereum => UserDataType::UserDataPrimaryAddressEthereum,
+            proto::Protocol::Solana => UserDataType::UserDataPrimaryAddressSolana,
+        };
+
+        // Check if this address is set as the primary address
+        let user_data_result = UserDataStore::get_user_data_by_fid_and_type(
+            &self.stores.user_data_store,
+            fid,
+            user_data_type,
+        );
+
+        // If we can't find the user data or there's an error, just return OK (nothing to revoke)
+        let user_data = match user_data_result {
+            Ok(data) => data,
+            Err(_) => return Ok(()),
+        };
+
+        // Check if the stored primary address matches the one being removed
+        let data_ref = match &user_data.data {
+            Some(data) => data,
+            None => return Ok(()),
+        };
+
+        let user_data_body = match &data_ref.body {
+            Some(Body::UserDataBody(body)) => body,
+            _ => return Ok(()),
+        };
+
+        // If the primary address matches the one being removed, revoke it
+        if user_data_body.value == parsed_address {
+            self.stores.user_data_store.revoke(&user_data, txn_batch)?;
```

Comment:
So, if you modify the store, we need to update the trie and emit the events. Otherwise the account state in the db will not reflect the merkle trie state.

We should capture the return value here and then return the list of events from `handle_delete_side_effects`, then similar to the `prune_messages` call, call update_trie for each event and then add these to the global list of events being returned from `replay_snapchain_txn`